### PR TITLE
FOAM_SLUG template variable addition

### PR DESCRIPTION
**Pull request by techCarpenter**
This is my first open source contribution. I'm excited to contribute and also eager to do things right so any feedback on this PR is appreciated!

---

This contribution addresses the discussion about [slugifying wikilinks](https://github.com/foambubble/foam/issues/494#issuecomment-907713254)

The `FOAM_TITLE_SLUG` variable will be available in foam templates. When the note title is entered with `Create New Note`, a slug is created by `github-slugger` that can then be used as part of the file name.

Example:

|Variable|Value|
|:-|:-|
|`FOAM_TITLE`|My new note|
|`FOAM_TITLE_SLUG`|my-new-note|

### CODE REVIEW

**Comment by chrisUsick** [comment]
I wanted to clarify something. Is it improper to use FOAM_TITLE in the file path? That, plus other recent fixes to slugs finally allowed for wikilinks to work well. I appreciate having the wikilink text having the capitalization and spaces. I have a lot of notes that use this format and would like to know if that’s becoming an unsupported pattern. 
Another thing I’m unclear on is whether FOAM_TITLE style wikiklinks affect publishing? It sounds like it might not work.

File: .all-contributorsrc [hunk 1]
```diff
@@ -31,6 +31,16 @@
         "doc"
       ]
     },
+    {
```
Comment (riccardoferretti) [inline .all-contributorsrc:34]: no need to add this part, I will do it independently via the allcontributors bot :)

File: docs/features/note-templates.md [hunk 2]
```diff
@@ -65,12 +67,12 @@ By using the `FOAM_DATE_` versions of the variables, the correct relative date w
 For example, given this daily note template (`.foam/templates/daily-note.md`):
 
 ```markdown
-# $FOAM_DATE_YEAR-$FOAM_DATE_MONTH-$FOAM_DATE_DATE
+# $FOAM_DATE_YEAR-$FOAM_DATE_MONTH-\$FOAM_DATE_DATE
```
Comment (riccardoferretti) [inline docs/features/note-templates.md:70]: why this?

File: docs/features/note-templates.md [hunk 3]
```diff
@@ -106,8 +108,9 @@ For example, `filepath` can be used to customize `.foam/templates/new-note.md`,
 # This will create the note in the "journal" subdirectory of the current workspace,
 # regardless of which file is the active file.
 foam_template:
-  filepath: 'journal/$FOAM_TITLE.md'
+  filepath: 'journal/$FOAM_TITLE_SLUG.md'
```
Comment (riccardoferretti) [inline docs/features/note-templates.md:111]: I am ok with the slug variable, but the Foam recommended way is still the other, so I prefer to keep FOAM_TITLE in the example

Comment (riccardoferretti) [reply docs/features/note-templates.md:111]: (this applies to all the other places where the FOAM_TITLE has been replaced)

File: packages/foam-vscode/src/services/variable-resolver.ts [hunk 4]
```diff
@@ -1,9 +1,11 @@
 import { findSelectionContent } from './editor';
 import { window } from 'vscode';
 import { UserCancelledOperation } from './errors';
+import slugger from 'github-slugger';
```
Comment (riccardoferretti) [inline packages/foam-vscode/src/services/variable-resolver.ts:4]: rather than importing the slugger, can you go through `utils/slug.ts`?

File: packages/foam-vscode/src/services/variable-resolver.ts [hunk 5]
```diff
@@ -55,6 +57,7 @@ export function findFoamVariables(templateText: string): string[] {
 
 export class Resolver {
   promises = new Map<string, Thenable<string>>();
+  foamTitle: string | null = null;
```
Comment (riccardoferretti) [inline packages/foam-vscode/src/services/variable-resolver.ts:60]: Why do we need to treat the title as a special case?
Could we, when resolving the slug, simply go through the resolution of the title?

**Review by riccardoferretti — COMMENTED** [review]
Thanks for your contribution @techCarpenter! I have left a few comments for you, let me know your thoughts

**Comment by techCarpenter** [comment]
@riccardoferretti Thank you for the notes, I'll address them!

File: docs/features/note-templates.md (hunk 3 above)
Comment (techCarpenter) [reply docs/features/note-templates.md:111]: No problem, I'll change the examples back to using the `FOAM_TITLE` variable 👌🏼

File: docs/features/note-templates.md (hunk 2 above)
Comment (techCarpenter) [reply docs/features/note-templates.md:70]: Oops, this must have been formatted automatically. I'll remove this backslash.

File: .all-contributorsrc (hunk 1 above)
Comment (techCarpenter) [reply .all-contributorsrc:34]: Ahh, good to know!

File: packages/foam-vscode/src/services/variable-resolver.ts (hunk 4 above)
Comment (techCarpenter) [reply packages/foam-vscode/src/services/variable-resolver.ts]: I made the changes, is that what you had in mind?

File: packages/foam-vscode/src/services/variable-resolver.ts (hunk 5 above)
Comment (techCarpenter) [reply packages/foam-vscode/src/services/variable-resolver.ts:60]: I tried using `resolveFoamTitle()` for the slug as well, but the input box would show twice or cancel the second one and not replace the variable `FOAM_TITLE_SLUG` in the template.

I can look into refactoring so the `foamTitle` property isn't required though.

**Comment by techCarpenter** [comment]
> I wanted to clarify something. Is it improper to use FOAM_TITLE in the file path? That, plus other recent fixes to slugs finally allowed for wikilinks to work well. I appreciate having the wikilink text having the capitalization and spaces. I have a lot of notes that use this format and would like to know if that’s becoming an unsupported pattern. Another thing I’m unclear on is whether FOAM_TITLE style wikiklinks affect publishing? It sounds like it might not work.

@chrisUsick I was told by @riccardoferretti that using FOAM_TITLE is the default way to do wikilinks in Foam. I suspect there could be issues when publishing files that have spaces in the filename.

**Comment by chrisUsick** [comment]
> > I wanted to clarify something. Is it improper to use FOAM_TITLE in the file path? That, plus other recent fixes to slugs finally allowed for wikilinks to work well. I appreciate having the wikilink text having the capitalization and spaces. I have a lot of notes that use this format and would like to know if that’s becoming an unsupported pattern. Another thing I’m unclear on is whether FOAM_TITLE style wikiklinks affect publishing? It sounds like it might not work.
> 
> @chrisUsick I was told by @riccardoferretti that using FOAM_TITLE is the default way to do wikilinks in Foam. I suspect there could be issues when publishing files that have spaces in the filename.



> > I wanted to clarify something. Is it improper to use FOAM_TITLE in the file path? That, plus other recent fixes to slugs finally allowed for wikilinks to work well. I appreciate having the wikilink text having the capitalization and spaces. I have a lot of notes that use this format and would like to know if that’s becoming an unsupported pattern. Another thing I’m unclear on is whether FOAM_TITLE style wikiklinks affect publishing? It sounds like it might not work.
> 
> @chrisUsick I was told by @riccardoferretti that using FOAM_TITLE is the default way to do wikilinks in Foam. I suspect there could be issues when publishing files that have spaces in the filename.



> > I wanted to clarify something. Is it improper to use FOAM_TITLE in the file path? That, plus other recent fixes to slugs finally allowed for wikilinks to work well. I appreciate having the wikilink text having the capitalization and spaces. I have a lot of notes that use this format and would like to know if that’s becoming an unsupported pattern. Another thing I’m unclear on is whether FOAM_TITLE style wikiklinks affect publishing? It sounds like it might not work.
> 
> @chrisUsick I was told by @riccardoferretti that using FOAM_TITLE is the default way to do wikilinks in Foam. I suspect there could be issues when publishing files that have spaces in the filename.

Ah, I see that in the comments. Thanks!

**Comment by memeplex** [comment]
The github slugger is not actually used, it seems to be a legacy part of Foam, I've opened a PR to remove the remnants (https://github.com/foambubble/foam/pull/872). Having a title is not required, but if your note has a title it's best to name the file after that title, or some transformation of it (but no transformation is enforced or preferred, therefore you have to do it manually, so again: better name your note after the title). Except that @riccardoferretti considers that some default transformation is desirable for templates.

**Comment by riccardoferretti** [comment]
The only constraints that Foam gives you is that wikilinks use the file name to find the resource.
As long as you are fine with this premise, you can use anything you want to name your files, whether the title, the slug, ... or something totally different (e.g. timestamp).

> Except that @riccardoferretti considers that some default transformation is desirable for templates

This is not about templates, per my comment above, you can pick whatever path you desire.

What would need a transformation might be the `[[wikilink key]]`, almost as an automatic alias (or, more precisely, a key normalization function). The way Foam works would actually support it fairly easily, but atm I am not inclined to have much magic happen there, as there are still a few wrinkles to iron before considering this option.

To summarize, sure you can have a `FOAM_TITLE = File with Slug` and via `FOAM_TITLE_SLUG` create your `/path/to/file-with-slug.md`, be aware that you will need to refer to such file as `[[file-with-slug]]` and not `[[File With Slug]]`.

**Comment by memeplex** [comment]
> This is not about templates, per my comment above, you can pick whatever path you desire.

What I meant is that when you create a template you have a title but not a slug, except when slug = template. But, sure, the same happens when you enter `[[slug]]` and ctrl-click on the placeholder link, if you want title = slug you will be fine with the default, otherwise you will have to edit the title. So, I was saying that except that you prefer some slug convention over the others, or are willing to provide a way to configure a programmable slugger, all conventions but title = slug will require extra manual intervention. Which is fine for me.

File: packages/foam-vscode/src/services/variable-resolver.ts (hunk 5 above)
Comment (riccardoferretti) [reply packages/foam-vscode/src/services/variable-resolver.ts:60]: have you tried `this.resolve('FOAM_TITLE')`?

**Comment by riccardoferretti** [comment]
relevant comment: https://github.com/foambubble/foam/pull/872#issuecomment-996170130

**Comment by memeplex** [comment]
Despite the outcome of the discussion ongoing in #872 I'd suggest to rename `FOAM_TITLE_SLUG` to `FOAM_SLUG`. slug is or may be used in other contexts, not necessarily derived from the title. Besides `FOAM_SLUG` is shorter.

**Comment by riccardoferretti** [comment]
What do you mean by "slug is or may be used in other contexts"?

**Comment by memeplex** [comment]
> What do you mean by "slug is or may be used in other contexts"?

More or less your open questions in https://github.com/foambubble/foam/pull/872#issuecomment-996232215. I mean, in this context it's clear that slug is something created from a title, but in other yet to be defined contexts that may not be the case. For example, in the ctrl-click over `[[a-placeholder]]` case maybe we end up creating a `# A placeholder` title from the slug. I believe a more neutral terminology of slug, like the slug being a feature of the note, instead of title slug, could be preferable.

File: packages/foam-vscode/src/services/variable-resolver.ts (hunk 5 above)
Comment (techCarpenter) [reply packages/foam-vscode/src/services/variable-resolver.ts:60]: I'll give `this.resolve('FOAM_TITLE')` a try this evening. Thanks for your patience!

File: docs/features/note-templates.md [hunk 6]
```diff
@@ -45,6 +45,7 @@ In addition, you can also use variables provided by Foam:
 | -------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | `FOAM_SELECTED_TEXT` | Foam will fill it with selected text when creating a new note, if any text is selected. Selected text will be replaced with a wikilink to the new note.                                                                     |
 | `FOAM_TITLE`         | The title of the note. If used, Foam will prompt you to enter a title for the note.                                                                                                                                         |
+| `FOAM_TITLE_SLUG`    | The sluggified title of the note. If used, Foam will prompt you to enter a title for the note unless `FOAM_TITLE` has already caused the prompt.                                                                            |
```
[the remark this answers is not in the document]
Comment (techCarpenter) [reply docs/features/note-templates.md:48]: `FOAM_SLUG` sounds great! It's more generic that way.

I'll add a note about the github slugging method.

File: packages/foam-vscode/src/services/variable-resolver.ts [hunk 7]
```diff
@@ -127,7 +129,10 @@ export class Resolver implements VariableResolver {
     } else if (!this.promises.has(name)) {
       switch (name) {
         case 'FOAM_TITLE':
-          this.promises.set(name, resolveFoamTitle());
+          this.promises.set(name, Promise.resolve(this.foamTitle));
+          break;
+        case 'FOAM_TITLE_SLUG':
+          this.promises.set(name, Promise.resolve(toSlug(this.foamTitle)));
```
[the remark this answers is not in the document]
Comment (techCarpenter) [reply packages/foam-vscode/src/services/variable-resolver.ts:135]: Thank you for the suggestion. Tomorrow I will make these changes so we can wrap up this PR 👌🏼

File: packages/foam-vscode/src/services/variable-resolver.ts [hunk 8]
```diff
@@ -26,7 +28,7 @@ const knownFoamVariables = new Set([
 
 export class Resolver implements VariableResolver {
   private promises = new Map<string, Promise<string | undefined>>();
-
+  private foamTitle = null;
```
Comment (riccardoferretti) [inline packages/foam-vscode/src/services/variable-resolver.ts:31]: I don't believe we need this anymore, am I missing something?

File: packages/foam-vscode/src/services/variable-resolver.ts [hunk 9]
```diff
@@ -251,19 +262,22 @@ export class Resolver implements VariableResolver {
...
-async function resolveFoamTitle() {
-  const title = await window.showInputBox({
-    prompt: `Enter a title for the new note`,
-    value: 'Title of my New Note',
-    validateInput: value =>
-      value.trim().length === 0 ? 'Please enter a title' : undefined,
-  });
-  if (title === undefined) {
-    throw new UserCancelledOperation();
+  async resolveFoamTitle() {
```
Comment (riccardoferretti) [inline packages/foam-vscode/src/services/variable-resolver.ts:266]: Should we be able to use the old version of this function at this point?

File: packages/foam-vscode/src/services/variable-resolver.ts (hunk 8 above)
Comment (techCarpenter) [reply packages/foam-vscode/src/services/variable-resolver.ts:31]: The function `resolveFoamTitle` is still assigning to that variable if I'm not mistaken.

File: packages/foam-vscode/src/services/variable-resolver.ts (hunk 9 above)
Comment (techCarpenter) [reply packages/foam-vscode/src/services/variable-resolver.ts:266]: Won't that require prompting the user twice if the value isn't stored somewhere?

Comment (riccardoferretti) [reply packages/foam-vscode/src/services/variable-resolver.ts:266]: because of the check in line 129 (`} else if (!this.promises.has(name)) {`) I don't believe that would be the case

File: packages/foam-vscode/src/services/variable-resolver.ts [hunk 10]
```diff
@@ -120,14 +122,23 @@ export class Resolver implements VariableResolver {
...
+          this.promises.set(
+            name,
+            Promise.resolve(await this.resolveFoamTitle())
+          );
+          break;
+        case 'FOAM_SLUG':
+          this.promises.set(
+            name,
+            Promise.resolve(toSlug(await this.resolveFoamTitle()))
+          );
```
Comment (riccardoferretti) [inline packages/foam-vscode/src/services/variable-resolver.ts:141]:
```suggestion
          this.promises.set(
            name,
            Promise.resolve(toSlug(await this.resolve('FOAM_TITLE')))
          );
```

I believe that if we use this way or getting the title we won't need to store it, because of the check at line 129, right before the switch statement.

Comment (techCarpenter) [reply packages/foam-vscode/src/services/variable-resolver.ts]: You were right so I've pushed those updates to the branch 🙌🏼

**Review by riccardoferretti — APPROVED** [review]
Looking good, thanks @techCarpenter!

**Comment by techCarpenter** [comment]
PR merged into master branch

**Comment by riccardoferretti** [comment]
@allcontributors add @techCarpenter for code
